beacon-chain/blockchain/service.go

```diff
@@ -235,14 +196,114 @@ func (c *ChainService) doesPoWBlockExist(block *types.Block) bool {
+// rule as canonical. This block is then persisted to storage.
+func (c *ChainService) updateHead(slotInterval <-chan time.Time) {
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-slotInterval:
+			c.currentSlot++
+
+			log.WithField("slotNumber", c.currentSlot).Info("New beacon slot")
+
+			// First, we check if there were any blocks processed in the previous slot.
+			// If there is, we fetch the first one from the DB.
+			if len(c.blocksPendingProcessing) == 0 {
+				continue
+			}
+
+			// Naive fork choice rule: we pick the first block we processed for the previous slot
+			// as canonical.
+			block, err := c.chain.getBlock(c.blocksPendingProcessing[0])
+			if err != nil {
+				log.Errorf("Could not get block: %v", err)
+				continue
+			}
+
+			h, err := block.Hash()
+			if err != nil {
+				log.Errorf("Could not hash incoming block: %v", err)
+				continue
+			}
+
+			log.Info("Applying fork choice rule")
+			aState := c.chain.ActiveState()
+			cState := c.chain.CrystallizedState()
+			isTransition := cState.IsCycleTransition(c.currentSlot - 1)
+
+			if isTransition {
+				cState, err = cState.NewStateRecalculations(aState, block)
+				if err != nil {
+					log.Errorf("Initialize new cycle transition failed: %v", err)
+					continue
+				}
+			}
+
+			parentBlock, err := c.chain.getBlock(block.ParentHash())
+			if err != nil {
+				log.Errorf("Failed to get parent of block 0x%x", h)
+				continue
+			}
+			aState, err = aState.CalculateNewActiveState(block, cState, parentBlock.SlotNumber())
+			if err != nil {
+				log.Errorf("Compute active state failed: %v", err)
+				continue
+			}
+
+			if err := c.chain.SetActiveState(aState); err != nil {
+				log.Errorf("Write active state to disk failed: %v", err)
+				continue
+			}
+
+			if err := c.chain.SetCrystallizedState(cState); err != nil {
+				log.Errorf("Write crystallized state to disk failed: %v", err)
+				continue
+			}
+
+			// Save canonical block hash with slot number to DB.
+			if err := c.chain.saveCanonicalSlotNumber(block.SlotNumber(), h); err != nil {
+				log.Errorf("Unable to save slot number to db: %v", err)
+				continue
+			}
+
+			// Save canonical block to DB.
+			if err := c.chain.saveCanonicalBlock(block); err != nil {
+				log.Errorf("Unable to save block to db: %v", err)
+				continue
+			}
+
+			log.WithField("blockHash", fmt.Sprintf("0x%x", h)).Info("Canonical block determined")
+
+			// We fire events that notify listeners of a new block (or crystallized state in
+			// the case of a state transition). This is useful for the beacon node's gRPC
+			// server to stream these events to beacon clients.
+			if isTransition {
+				c.canonicalCrystallizedStateFeed.Send(cState)
+			}
+			c.canonicalBlockFeed.Send(block)
+
+			// Clear the blocks pending processing, mutex lock for thread safety
+			// in updating this slice.
+			c.lock.Lock()
+			c.blocksPendingProcessing = [][32]byte{}
```

Comment:
The lock needs to be acquired [here](https://github.com/prysmaticlabs/prysm/pull/474/files#diff-7f087f690e8ec9178851c9b6844cb529R216) since it's being accessed here.

Reply:
An alternative to using locks would be to communicate using channels. The block processing goroutine would be sending these blocks through a channel, and this goroutine would have a select case that processes these blocks into an array. A second select case triggers every `slotDuration` seconds to run this function. Since there's no shared state, there's no need for locks.

This second option might be better since there's no locks but I'm okay with leaving this as is.